pkg/resource/deploy/step.go

```diff
@@ -391,6 +397,103 @@ func (s *ReplaceStep) Apply(preview bool) (resource.Status, error) {
+// target of a RegisterResource in the next plan, a CreateReplacement step will be issued to indicate the transition
+// from external to owned. If a URN that was previously not marked "External" is the target of a ReadResource in the
+// next plan, a ReadReplacement step will be issued to indicate the transition from owned to external.
+type ReadStep struct {
+	plan      *Plan             // the plan that produced this read
+	event     ReadResourceEvent // the event that should be signaled upon completion
+	old       *resource.State   // the old resource state, if one exists for this urn
+	new       *resource.State   // the new resource state, to be used to query the provider
+	replacing bool              // whether or not the new resource is replacing the old resource
+}
+
+// NewReadStep creates a new Read step.
+func NewReadStep(plan *Plan, event ReadResourceEvent, old *resource.State, new *resource.State) Step {
+	contract.Assert(new != nil)
+	contract.Assertf(new.External, "target of Read step must be marked External")
+	contract.Assertf(new.Custom, "target of Read step must be Custom")
+
+	// If Old was given, it can't be an external resource.
+	contract.Assert(old == nil || old.External)
+	return &ReadStep{
+		plan:      plan,
+		event:     event,
+		old:       old,
+		new:       new,
+		replacing: false,
+	}
+}
+
+// NewReadReplacementStep creates a new Read step with the `replacing` flag set. When executed,
+// it will pend deletion of the "old" resource, which must not be an external resource.
+func NewReadReplacementStep(plan *Plan, event ReadResourceEvent, old *resource.State, new *resource.State) Step {
+	contract.Assert(new != nil)
+	contract.Assertf(new.External, "target of ReadReplacement step must be marked External")
+	contract.Assertf(new.Custom, "target of ReadReplacement step must be Custom")
+	contract.Assert(old != nil)
+	contract.Assertf(!old.External, "old target of ReadReplacement step must not be External")
+	return &ReadStep{
+		plan:      plan,
+		event:     event,
+		old:       old,
+		new:       new,
+		replacing: true,
+	}
+}
+
+func (s *ReadStep) Op() StepOp {
+	if s.replacing {
+		return OpReadReplacement
+	}
+
+	return OpRead
+}
+
+func (s *ReadStep) Plan() *Plan          { return s.plan }
+func (s *ReadStep) Type() tokens.Type    { return s.new.Type }
+func (s *ReadStep) URN() resource.URN    { return s.new.URN }
+func (s *ReadStep) Old() *resource.State { return s.old }
+func (s *ReadStep) New() *resource.State { return s.new }
+func (s *ReadStep) Res() *resource.State { return s.new }
+func (s *ReadStep) Logical() bool        { return !s.replacing }
+
+func (s *ReadStep) Apply(preview bool) (resource.Status, error) {
+	urn := s.new.URN
+	id := s.new.ID
+
+	// Unlike most steps, Read steps run during previews. The only time
+	// we can't run is if the ID we are given is unknown.
+	if id != plugin.UnknownStringValue {
```

Comment:
This might need to be `if id != ""`. At least in the case of `RegisterResource`, the `id` isn't a typical `PropertyValue`, it's just a string.